### PR TITLE
CPU Layout Rewrite

### PR DESCRIPTION
**Pull request by junrushao**
This PR is intended to be split into the following commits:
- [Relay][Op] MetaSchedule layout in TypeRel
- [TOPI][Relay] New Op: MetaScheduleLayoutRewrite
- [TOPI] Layout Rewriting in TE
- [Relay][Pass] Meta-Schedule-Layout-Rewrite
- [OpStrategy] Support MetaSchedule Layout
- [MetaSchedule] Introduce ArgInfo::FromEntryFunc
- [TIR][Pass] Remove-Weight-Layout-Rewrite-Block
- [MetaSchedule] Postproc: Rewrite-Layout

### CODE REVIEW

File: include/tvm/topi/transform.h [hunk 1]
```diff
@@ -1693,6 +1694,21 @@ inline Tensor auto_scheduler_layout_transform(const Tensor& src, const String& s
...
+ * \param src the source input.
+ * \param index_map The TIR IndexMap
+ * \param name output tensor name.
+ * \param tag output tensor tag.
+ * \return A tensor. The layout transformation method
+ */
+inline Tensor meta_schedule_layout_transform(const Tensor& src, const tir::IndexMap& index_map,
+                                             const String name = "T_meta_schedule_layout_trans",
+                                             const String tag = kInjective) {
+  // TODO
```
Comment (junrushao) [inline include/tvm/topi/transform.h]: @jinhongyii I will need help on this

Comment (jinhongyii) [reply include/tvm/topi/transform.h]: Doesn't InverseAffineIterMap work?

File: src/relay/op/tensor/transform.cc [hunk 2]
```diff
@@ -3348,6 +3348,43 @@ RELAY_REGISTER_OP("auto_scheduler_layout_transform")
...
+                                                     const Array<te::Tensor>& inputs,
+                                                     const Type& out_type) {
+  const auto* param = attrs.as<MetaScheduleLayoutTransformAttrs>();
+  CHECK(param != nullptr);
+  return Array<te::Tensor>{topi::meta_schedule_layout_transform(inputs[0], param->index_map)};
+}
+
+bool MetaScheduleLayoutTransformRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
+                                    const TypeReporter& reporter) {
+  // TODO
```
Comment (junrushao) [inline src/relay/op/tensor/transform.cc]: @jinhongyii I will need your help on this too

Comment (jinhongyii) [reply src/relay/op/tensor/transform.cc]: what is this function used for?

File: include/tvm/topi/transform.h (hunk 1 above)
Comment (junrushao) [reply include/tvm/topi/transform.h]: To this end, I was actually examining the code here but didn't found `InverseAffineIterMap` used anywhere in TransformLayout: https://github.com/junrushao1994/tvm/blob/65071978d01af5ba2ca9721ba51125fe46d6887d/src/tir/schedule/primitive/layout_transformation.cc#L135. am a bit confused now

File: src/relay/op/tensor/transform.cc (hunk 2 above)
Comment (junrushao) [reply src/relay/op/tensor/transform.cc]: if we address the previous problem (https://github.com/junrushao1994/tvm/pull/48#discussion_r901076699), then this one should be easy. let's focus on the previous one first

File: include/tvm/topi/transform.h (hunk 1 above)
Comment (jinhongyii) [reply include/tvm/topi/transform.h]: why do you expect InverseAffineIterMap to be used in TransformLayout
?

File: include/tvm/topi/transform.h [hunk 3]
```diff
@@ -1693,6 +1694,59 @@ inline Tensor auto_scheduler_layout_transform(const Tensor& src, const String& s
...
+    iter_domain.push_back(Range::FromMinExtent(make_zero(e->dtype), e));
+  }
+  Array<PrimExpr> post_transform_shape = index_map->MapShape(src->shape);
+  return compute(
+      post_transform_shape,
+      [src, inv = index_map.Inverse(iter_domain)](const Array<Var>& indices) -> PrimExpr {
+        return src(inv->MapIndices(Array<PrimExpr>{indices.begin(), indices.end()}));
+      },
+      name, tag);
+}
```
Comment (junrushao) [inline include/tvm/topi/transform.h:1748]: @jinhongyii would you mind taking a look at the impl and make sure it's consistent with your design?

Comment (jinhongyii) [reply include/tvm/topi/transform.h:1748]: yes, it's consistent

**Pull request closed without merging**